File: python/day_09.py
import itertools
from tqdm import tqdm


def play(num_players, last_marble):
    players = itertools.cycle(range(num_players))
    scores = [0 for _ in range(num_players)]

    current_player = next(players)
    marbles = [0]
    current_index = 0

    for marble in tqdm(range(1, last_marble + 1)):
        current_player = next(players)

        if marble % 23 == 0:
            remove_index = (current_index - 7) % len(marbles)
            scores[current_player] += marble + marbles.pop(remove_index)
            current_index = remove_index
        else:
            next_index = (current_index + 2) % len(marbles)
            marbles.insert(next_index, marble)
            current_index = next_index

    return max(scores)


def part_1():
    # return play(9, 25)
    # return play(10, 1618)
    return play(411, 71170)


def part_2():
    return play(411, 71170 * 100)


if __name__ == '__main__':
    print(f'Part 1: {part_1()}')
    print(f'Part 2: {part_2()}')

File: ruby/day_09.rb
# frozen_string_literal: true

def play(num_players, num_marbles)
  marbles = [0]
  scores = Hash.new(0)
  current_marble_idx = 0
  current_player = 0

  (1...num_marbles).each do |marble_number|
    puts 100 * marble_number / num_marbles.to_f if (marble_number % 10000).zero?

    if (marble_number % 23).zero?
      current_marble_idx = (current_marble_idx - 7) % marbles.length
      scores[current_player] += marble_number + marbles.delete_at(current_marble_idx)
    else
      current_marble_idx = (current_marble_idx + 2) % marbles.length
      marbles.insert(current_marble_idx, marble_number)
    end

    current_player = (current_player + 1) % num_players
  end

  scores.values.max
end

def part_one
  play(411, 71_170)
end

def part_two
  play(411, 71_170 * 100)
end

if $PROGRAM_NAME == __FILE__
  puts 'https://adventofcode.com/2018/day/9'

  puts "Part One: #{part_one}"
  puts "Part Two: #{part_two}"
end
